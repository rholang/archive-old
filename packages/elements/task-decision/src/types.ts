import { ServiceConfig } from '@atlaskit/util-service-support';

export type DecisionState = 'DECIDED';
export type DecisionStatus = 'CREATED';
export type TaskState = 'TODO' | 'DONE';
export type Cursor = string;

export type DecisionType = 'DECISION';
export type TaskType = 'TASK';

export interface ContentRef {
  (ref: HTMLElement | null): void;
}

export interface ObjectKey {
  localId: string;
  containerAri?: string;
  objectAri: string;
}

export interface BaseItem<S> extends ObjectKey {
  state: S;
  lastUpdateDate: Date;
  type: DecisionType | TaskType;
}

export interface ServiceDecision {
  creationDate?: string;
  creatorId?: UserId;
  lastUpdaterId?: UserId;
  lastUpdateDate: string;
  localId: string;
  objectAri: string;
  participants?: UserId[];
  state?: DecisionState;
  status: DecisionStatus;
  type: DecisionType;
}

export interface Meta {
  cursor?: string;
}

export interface ServiceDecisionResponse {
  decisions: ServiceDecision[];
  meta: Meta;
}

export type ServiceItem = ServiceDecision | ServiceTask;

export interface ServiceTaskState {
  lastUpdateDate: string;
  localId: string;
  objectAri: string;
  state: TaskState;
}

export interface Decision extends BaseItem<DecisionState> {
  creationDate?: Date;
  creator?: UserId;
  lastUpdater?: UserId;
  lastUpdateDate: Date;
  participants?: UserId[];
  status: DecisionStatus;
  type: DecisionType;
}

export type Item = Decision | Task;

export type UserId = string;

export interface ServiceTask {
  creationDate?: string;
  creatorId?: UserId;
  lastUpdaterId?: UserId;
  lastUpdateDate: string;
  localId: string;
  objectAri: string;
  parentLocalId?: string;
  participants?: UserId[];
  position: number;
  state: TaskState;
  type: TaskType;
}

export interface Task extends BaseItem<TaskState> {
  creationDate?: Date;
  creator?: UserId;
  lastUpdater?: UserId;
  lastUpdateDate: Date;
  parentLocalId?: string;
  participants?: UserId[];
  position?: number;
  type: TaskType;
}

export type Handler = (state: TaskState | DecisionState) => void;

export type RecentUpdatesId = string;

export interface RecentUpdateContext {
  objectAri: string;
  localId?: string;
}

/**
 * A subscriber interface that can be called back if there are new decisions/tasks/items
 * available as the result of an external change.
 */
export interface RecentUpdatesListener {
  /**
   * An id that can be used to unsubscribe
   */
  id(id: RecentUpdatesId): void;

  /**
   * Indicates there are recent updates, and the listener should refresh
   * the latest items from the TaskDecisionProvider.
   *
   * There will be a number of retries until expectedLocalId, if passed.
   *
   * @param updateContext Recent update context
   */
  recentUpdates(updateContext: RecentUpdateContext): void;
}

export interface TaskDecisionResourceConfig extends ServiceConfig {
  pubSubClient?: PubSubClient;

  /**
   * Indicates if initial state for an action or decision is should be cached,
   * from the content, i.e. was originally hydrated from the service initially,
   * and so should be considered up to date.
   *
   * Will stop the initiation of the hydration from the service the first
   * time an action or decision is seen.
   *
   * If false the state will always be hydrated from the service on first view.
   */
  disableServiceHydration?: boolean;
}

export interface TaskDecisionProvider {
  unsubscribeRecentUpdates(id: RecentUpdatesId): void;
  notifyRecentUpdates(updateContext: RecentUpdateContext): void;

  // Tasks
  toggleTask(objectKey: ObjectKey, state: TaskState): Promise<TaskState>;
  subscribe(
    objectKey: ObjectKey,
    handler: Handler,
    item?: BaseItem<TaskState | DecisionState>,
  ): void;
  unsubscribe(objectKey: ObjectKey, handler: Handler): void;
}

/**
 * Same as RendererContext in editor-core (don't want an direct dep though)
 */
export interface RendererContext {
  objectAri: string;
  containerAri?: string;
}

export interface RenderDocument {
  (document: any, rendererContext?: RendererContext): JSX.Element;
}

export interface OnUpdate<T> {
  (allDecisions: T[], newDecisions: T[]): void;
}

export type Appearance = 'inline';

/**
 * Same as PubSub client types (don't want a direct dep though)
 */

export type ARI = string;
export type AVI = string;

export interface PubSubOnEvent<T = any> {
  (event: string, data: T): void;
}

export interface PubSubClient {
  on(eventAvi: string, listener: PubSubOnEvent): PubSubClient;

  off(eventAvi: string, listener: PubSubOnEvent): PubSubClient;

  join(aris: ARI[]): Promise<PubSubClient>;

  leave(aris: ARI[]): Promise<PubSubClient>;
}

export enum PubSubSpecialEventType {
  ERROR = 'ERROR',
  CONNECTED = 'CONNECTED',
  RECONNECT = 'RECONNECT',
}
