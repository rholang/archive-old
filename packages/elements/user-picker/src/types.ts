import * as React from 'react';
import { EmailValidator } from './components/emailValidation';

export type UserPickerProps = {
  /**
   * Used to configure additional information regarding where the
   * user picker has been mounted.
   *
   * The purpose is to give more context as to where user picker events
   * are being fired from, as the current data may not uniquely identify
   * which field is the source.
   *
   * The value will be passed as a data attribute for analytics.
   * Examples include "assignee", "watchers" and "share".
   *
   * A second usage for the fieldId is for server side rendering (SSR) where it must be a unique id per UserPicker
   * instance contained in the serialized SSR content. E.g. a SPA page rendered through SSR that has multiple user pickers.
   *
   * fieldId can be set to null if the integrator is not listening
   * for the analytic events or does not care about SSR.
   */
  fieldId: string | null;
  /** List of users or teams to be used as options by the user picker. */
  options?: OptionData[];
  /** Width of the user picker field. It can be the amount of pixels as numbers or a string with the percentage. */
  width?: number | string;
  /** Sets the minimum width for the menu. If not set, menu will always have the same width of the field. */
  menuMinWidth?: number;
  /**
   * Function used to load options asynchronously.
   * accepts two optional params:
   * searchText?: optional text to filter results
   * sessionId?: user picker session identifier, used to track success metric for users providers
   */
  loadOptions?: LoadOptions;
  /** Callback for value change events fired whenever a selection is inserted or removed. */
  onChange?: OnChange;
  /** To enable multi user picker. */
  isMulti?: boolean;
  /** Input text value. */
  search?: string;
  /** Anchor for the user picker popup. */
  anchor?: React.ComponentType<any>;
  /** Controls if user picker menu is open or not. If not provided, UserPicker will control menu state internally. */
  open?: boolean;
  /** Show the loading indicator. */
  isLoading?: boolean;
  /** Callback for search input text change events. */
  onInputChange?: OnInputChange;
  /** Callback for when a selection is made. */
  onSelection?: OnOption;
  /** Callback for when the field gains focus. */
  onFocus?: OnPicker;
  /** Callback for when the field loses focus. */
  onBlur?: OnPicker;
  /** Callback for when the value/s in the picker is cleared. */
  onClear?: OnPicker;
  /** Callback that is triggered when popup picker is closed */
  onClose?: OnPicker;
  /** Appearance of the user picker. */
  appearance?: Appearance;
  /** Display the picker with a subtle style. */
  subtle?: boolean;
  /** Default value for the field to be used on initial render. */
  defaultValue?: Value;
  /** Placeholder text to be shown when there is no value in the field. */
  placeholder?: React.ReactNode;
  /** Message to encourage the user to add more items to user picker. */
  addMoreMessage?: string;
  /** Message to be shown when the menu is open but no options are provided.
   * If message is null, no message will be displayed.
   * If message is undefined, default message will be displayed.
   */
  noOptionsMessage?:
    | ((value: { inputValue: string }) => string | null)
    | string
    | null;
  /** Controls if the user picker has a value or not. If not provided, UserPicker will control the value internally. */
  value?: Value;
  /** Disable all interactions with the picker, putting it in a read-only state. */
  isDisabled?: boolean;
  /** Display a remove button on the single picker. True by default. */
  isClearable?: boolean;
  /** Optional tooltip to display on hover over the clear indicator. */
  clearValueLabel?: string;
  /** Whether the menu should use a portal, and where it should attach. */
  menuPortalTarget?: HTMLElement;
  /** Whether the user is allowed to enter emails as a value. */
  allowEmail?: boolean;
  /** Email option label */
  emailLabel?: string;
  /** Whether to disable interaction with the input */
  disableInput?: boolean;
  /** Override default email validation function. */
  isValidEmail?: EmailValidator;
  /** Override the internal behaviour to automatically focus the control when the picker is open */
  autoFocus?: boolean;
  /** The maximum number options to be displayed in the dropdown menu during any state of search. The value should be non-negative. */
  maxOptions?: number;
  /** Allows clicking on a label with the same id to open user picker. */
  inputId?: string;
};

export type PopupUserPickerProps = UserPickerProps & {
  /** Whether to use the popup version of the single picker */
  target: Target;
  /** Optional title assigned to popup picker */
  popupTitle?: string;
  /**
   *  Reference to the HTMLElement that should be used as a boundary for the popup.
   *  Viewport is used by default.
   */
  boundariesElement?: HTMLElement;
};

export type UserPickerState = {
  options: OptionData[];
  value?: AtlaskitSelectValue;
  inflightRequest: number;
  count: number;
  hoveringClearIndicator: boolean;
  menuIsOpen: boolean;
  inputValue: string;
  resolving: boolean;
};

export interface HighlightRange {
  start: number;
  end: number;
}

export interface UserHighlight {
  name: HighlightRange[];
  publicName: HighlightRange[];
}

export interface TeamHighlight {
  name: HighlightRange[];
  description?: HighlightRange[];
}

export interface GroupHighlight {
  name: HighlightRange[];
}

export interface OptionData {
  id: string;
  name: string;
  type?: 'user' | 'team' | 'email' | 'group';
  fixed?: boolean;
}

export const UserType = 'user';

export interface User extends OptionData {
  avatarUrl?: string;
  publicName?: string;
  highlight?: UserHighlight;
  byline?: string;
  type?: 'user';
}

export const TeamType = 'team';

export interface Team extends OptionData {
  avatarUrl?: string;
  description?: string;
  memberCount?: number;
  includesYou?: boolean;
  highlight?: TeamHighlight;
  type: 'team';
}

export const GroupType = 'group';

export interface Group extends OptionData {
  highlight?: GroupHighlight;
  type: 'group';
}

export type Value = OptionData | OptionData[] | null | undefined;
export const EmailType = 'email';

export interface Email extends OptionData {
  type: 'email';
}

export type ActionTypes =
  | 'select-option'
  | 'deselect-option'
  | 'remove-value'
  | 'pop-value'
  | 'set-value'
  | 'clear'
  | 'create-option';

export type OnChange = (value: Value, action: ActionTypes) => void;

export type OnInputChange = (query?: string, sessionId?: string) => void;

export type OnPicker = (sessionId?: string) => void;

export type OnOption = (value: Value, sessionId?: string) => void;

export type Option = {
  label: string;
  value: string;
  data: OptionData;
};

export interface LoadOptions {
  (searchText?: string, sessionId?: string):
    | Promisable<OptionData | OptionData[]>
    | Iterable<
        Promisable<OptionData[] | OptionData> | OptionData | OptionData[]
      >;
}

export type Promisable<T> = T | PromiseLike<T>;

export type InputActionTypes =
  | 'set-value'
  | 'input-change'
  | 'input-blur'
  | 'menu-close';

export type AtlaskitSelectValue = Option | Array<Option> | null | undefined;

export type AtlasKitSelectChange = (
  value: AtlaskitSelectValue,
  extraInfo: {
    removedValue?: Option;
    option?: Option;
    action: ActionTypes;
  },
) => void;

export type Appearance = 'normal' | 'compact';

export type Target = (withRef: { ref: any }) => any;
