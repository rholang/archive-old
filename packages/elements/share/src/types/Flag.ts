export const OBJECT_SHARED = 'object-shared';

export const ADMIN_NOTIFIED = 'admin-notified';

export type FlagType = 'object-shared' | 'admin-notified';

export type MessageDescriptor = {
  id: string;
  description: string;
  defaultMessage: string;
};

export type Flag = {
  appearance: 'success';
  title: MessageDescriptor;
  type: FlagType;
};
