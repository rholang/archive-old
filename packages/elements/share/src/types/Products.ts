/** Product ID (Canonical ID) in ARI of the share request */
export type ProductId =
  | 'bitbucket'
  | 'confluence'
  | 'jira-core'
  | 'jira-servicedesk'
  | 'jira-software'
  | 'trello';
