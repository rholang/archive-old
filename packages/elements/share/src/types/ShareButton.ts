export type ShareButtonStyle = 'icon-only' | 'icon-with-text' | 'text-only';
export type TooltipPosition = 'top' | 'right' | 'bottom' | 'left' | 'mouse';
