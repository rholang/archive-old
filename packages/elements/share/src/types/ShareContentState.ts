import { OptionData } from '@atlaskit/user-picker';
import { Comment } from './ShareEntities';
import { User } from './User';

export type ShareContentState = {
  users: User[];
  comment?: Comment;
};

export type ShareError = {
  message: string;
};

export type DialogContentState = {
  users: OptionData[];
  comment?: Comment;
};
