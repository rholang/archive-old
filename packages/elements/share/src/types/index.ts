export {
  ConfigResponse,
  ConfigResponseMode,
  ShareClient,
  ShareRequest,
  ShareResponse,
} from '../clients/ShareServiceClient';
export * from './form';
export * from './Flag';
export * from './OriginTracing';
export * from './Products';
export * from './ShareButton';
export {
  DialogContentState,
  ShareContentState,
  ShareError,
} from './ShareContentState';
export * from './ShareDialogWithTrigger';
export { Comment, Content, MetaData } from './ShareEntities';
export { User, UserWithEmail, UserWithId } from './User';

export type KeysOfType<T, TProp> = {
  [P in keyof T]: T[P] extends TProp ? P : never
}[keyof T];
