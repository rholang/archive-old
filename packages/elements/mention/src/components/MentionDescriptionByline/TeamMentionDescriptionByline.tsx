import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { DescriptionBylineProps } from './types';
import { DescriptionBylineStyle } from './styles';
import { messages } from '../i18n';

export default class TeamMentionDescriptionByline extends React.PureComponent<
  DescriptionBylineProps
> {
  private renderByline = (memberCount: number, includesYou: boolean) => {
    if (includesYou) {
      if (memberCount > 50) {
        return this.getBylineComponent(
          <FormattedMessage {...messages.plus50MembersWithYou} />,
        );
      }

      return this.getBylineComponent(
        <FormattedMessage
          {...messages.memberCountWithYou}
          values={{
            0: memberCount,
          }}
        />,
      );
    } else {
      if (memberCount > 50) {
        return this.getBylineComponent(
          <FormattedMessage {...messages.plus50MembersWithoutYou} />,
        );
      }

      return this.getBylineComponent(
        <FormattedMessage
          {...messages.memberCountWithoutYou}
          values={{
            0: memberCount,
          }}
        />,
      );
    }
  };

  private getBylineComponent = (message: JSX.Element) => (
    <DescriptionBylineStyle>{message}</DescriptionBylineStyle>
  );

  render() {
    const { context } = this.props.mention;

    if (!context) {
      return null;
    }

    const includesYou = context.includesYou;
    const memberCount = context.memberCount;

    return this.renderByline(memberCount, includesYou);
  }
}
