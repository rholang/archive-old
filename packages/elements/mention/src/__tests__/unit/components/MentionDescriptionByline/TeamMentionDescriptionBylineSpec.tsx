import { render } from 'enzyme';
import * as React from 'react';
import TeamMentionDescriptionByline from '../../../../components/MentionDescriptionByline';
import { MentionDescription } from '../../../..//types';
import { teamMention } from './_commonData';

const renderByline = (teamData: MentionDescription) => {
  return render(<TeamMentionDescriptionByline mention={teamData} />);
};

const buildTeamData = (memberCount: number, includesYou: boolean) => ({
  ...teamMention,
  context: {
    members: [
      {
        id: 'user-1234',
        name: 'Test User',
      },
    ],
    ...teamMention.context,
    includesYou: includesYou,
    memberCount: memberCount,
    teamLink: '/wiki/people/team/12345',
  },
});

describe('Team mention description', () => {
  it('should render Team Mention description component with less than 50 members, includes you', () => {
    const component = renderByline(buildTeamData(5, true));
    expect(component).toMatchSnapshot();
  });

  it('should render Team Mention description component with more than 50 members, includes you', () => {
    const component = renderByline(buildTeamData(55, true));
    expect(component).toMatchSnapshot();
  });

  it('should render Team Mention description component with less than 50 members, not including you', () => {
    const component = renderByline(buildTeamData(4, false));
    expect(component).toMatchSnapshot();
  });

  it('should render Team Mention description component with more than 50 members, not including you', () => {
    const component = renderByline(buildTeamData(100, false));
    expect(component).toMatchSnapshot();
  });

  it('should not render Team Mention description component if context is not given', () => {
    const newMention = { ...teamMention };
    delete newMention.context;

    const component = renderByline(newMention);
    expect(component).toMatchSnapshot();
  });
});
