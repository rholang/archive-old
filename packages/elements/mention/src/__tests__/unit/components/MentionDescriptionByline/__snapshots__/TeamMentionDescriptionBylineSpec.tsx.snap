// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team mention description should not render Team Mention description component if context is not given 1`] = `null`;

exports[`Team mention description should render Team Mention description component with less than 50 members, includes you 1`] = `
.c0 {
  color: #7A869A;
  font-size: 12px;
  margin-top: 2px;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  class="c0"
>
  <span>
    5 members, including you
  </span>
</span>
`;

exports[`Team mention description should render Team Mention description component with less than 50 members, not including you 1`] = `
.c0 {
  color: #7A869A;
  font-size: 12px;
  margin-top: 2px;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  class="c0"
>
  <span>
    4 members
  </span>
</span>
`;

exports[`Team mention description should render Team Mention description component with more than 50 members, includes you 1`] = `
.c0 {
  color: #7A869A;
  font-size: 12px;
  margin-top: 2px;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  class="c0"
>
  <span>
    50+ members, including you
  </span>
</span>
`;

exports[`Team mention description should render Team Mention description component with more than 50 members, not including you 1`] = `
.c0 {
  color: #7A869A;
  font-size: 12px;
  margin-top: 2px;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  class="c0"
>
  <span>
    50+ members
  </span>
</span>
`;
